Make ext4 rootfs great again

rouming opened the pull request:
This set of patches fixes the ext4 rootfs generation. Main problem was in the image size difference between squashfs image and ext4 which leads to an incorrect GPT table. The solution is to calculate resulting image dynamically. 

Patch makes creation of the ext4 rootfs possible and also has a size reduction for the installer.raw images:

```    
                      before      after
     installer.raw     592Mb      296Mb
```

Unfortunately for the rootfs we have a hardcoded partition size equal to 512Mb, which can't be changed, otherwise upgrade logic fail when we try to fit bigger rootfs into smaller one partition. This means that upgrade from squashfs image to ext4 is currently impossible. But who cares.

Now live image can be generated as following:
```
  make ROOTFS_FORMAT=ext4 live
```

My motivation to generate ext4 rootfs is to have possibility to persist changes on eve while running. 
Another great feature becomes possible is to "patch" resulting qemu image with changed files and not generate rootfs from scratch, which can take minutes of CPU time.

This patch changes the pkg/mkverification-raw-efi/make-raw, which almost a copy of the original one. I kindly ask @jsfakian to help me to review and to test the resulting verification images, that those work and nothing is broken because of my changes.

cc: @mikem-zed

jsfakian commented on Makefile:
```diff
@@ -737,7 +739,7 @@ $(LIVE).parallels: $(LIVE).raw
 $(VERIFICATION).raw: $(BOOT_PART) $(EFI_PART) $(ROOTFS_IMG) $(INITRD_IMG) $(VERIFICATION_IMG) $(CONFIG_IMG) $(PERSIST_IMG) $(BSP_IMX_PART) | $(VERIFICATION)
 	./tools/prepare-platform.sh "$(PLATFORM)" "$(BUILD_DIR)" "$(VERIFICATION)" || :
-	./tools/makeflash.sh "mkverification-raw-efi" -C 850 $| $@ "conf_win verification inventory_win"
+	./tools/makeflash.sh "mkverification-raw-efi" -C $| $@ "conf_win verification inventory_win"
```
I like these changes.

jsfakian commented on pkg/mkimage-raw-efi/make-raw:
```diff
@@ -24,8 +24,9 @@
-#   [-r] <img> [part1...]
+#   [-C] [-r] <img> [part1...]
 #
+# -C          recreate disk image
```
What does that mean?

jsfakian commented on pkg/mkimage-raw-efi/make-raw:
```diff
@@ -70,21 +72,43 @@ IMGB_UUID=ad6871ee-31f9-4cf3-9e09-6f7a25c30053
+IMX8_BLOB=/parts/imx8-flash.bin
+# IMX8 configuration file with flashing parameters
+IMX8_CONF=/parts/imx8-flash.conf
+
```
So with the current change, instead of statically define sizes for each image we calculate it?

rouming replied:
yes. not for everything, mostly rootfs

jsfakian commented on pkg/mkimage-raw-efi/make-raw:
```diff
@@ -101,6 +125,85 @@ The operation has completed successfully.' | tr '\012' '|')$^" || :
+  SIZE=$((SIZE + "$(file_size $INITRD_IMG)"))
+  SIZE=$((SIZE + "$(file_size $ROOTFS_IMG)"))
+  SIZE=$((SIZE + "$(file_size $INSTALLER_IMG)"))
+  SIZE=$((SIZE + "$(file_size $INSTALLER_GRUB_CFG)"))
```
Wouldn't a for be better here?

rouming replied:
for over what?

jsfakian commented on pkg/mkrootfs-ext4/make-rootfs:
```diff
@@ -41,11 +41,11 @@ IMGFILE=/rootfs.img
   ROOTFS_PART_SIZE=$(( $ROOTFS_PART_BLOCKS * $ROOTFS_BLOCKSZ ))
-  ROOTFS_PART_SIZE_KB=$(( ( ($ROOTFS_PART_SIZE + 1024) / 1024 ) / 1024 * 1024 ))
-  ROOTFS_PART_SECTORS=$(( $ROOTFS_PART_SIZE_KB * 2 ))
+  ROOTFS_PART_SIZE_KB=$(((ROOTFS_PART_SIZE + 1023) / 1024))
```
`ROOTFS_PART_SIZE_KB=$(((ROOTFS_PART_SIZE + 1023) / 1024))` produces different results compared to `ROOTFS_PART_SIZE_KB=$(( ( ($ROOTFS_PART_SIZE + 1024) / 1024 ) / 1024 * 1024 ))`. Is this ok?

jsfakian replied:
```
#!/bin/bash

calc() {
        x=$1
        echo $(( ( (x + 1024) / 1024 ) / 1024 * 1024 ))
}

calc2() {
        x=$1
        echo $(((x + 1023) / 1024))
}

calc 1025*1024
calc2 1025*1024
```

rouming replied:
yes, different. because old implementation bogously reduces size, but we expect round up. when we provide exactly 1025kb, why the size of the resulting image is reduced to 1024kb?

deitch commented:
Does this change anything in the build docs [here](https://github.com/lf-edge/eve/blob/master/docs/BUILD.md)? Both internal as to how build is done and the UX (which we care about more)?

deitch reviewed (approved):
I dislike the near-duplication of `make-raw` in both `mkimage-raw-efi` and `mkverification-raw-efi`, but that wasn't introduced here, so not your job to fix here. :-)

In any case, as long as this all works well and is tested, looks good for what it does. I like that it calculates size dynamically.

deitch commented:
> Another great feature becomes possible is to "patch" resulting qemu image with changed files and not generate rootfs from scratch, which can take minutes of CPU time

How does that work?

eriknordmark reviewed (commented):
What's the impact on the current squashfs installer and live image from the dynamic size?
I can't tell whether the partition sizes will be dynamic or fixed with these changes.

eriknordmark commented:
> What's the impact on the current squashfs installer and live image from the dynamic size? I can't tell whether the partition sizes will be dynamic or fixed with these changes.

Answering my own question, the partition size (at least in the live image) is the size of the image. 
linuxkit-525400123456:~# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
sda      8:0    0   64G  0 disk 
├─sda1   8:1    0   36M  0 part 
├─sda2   8:2    0  232M  0 part /
├─sda3   8:3    0  232M  0 part 
├─sda4   8:4    0    1M  0 part 
└─sda9   8:9    0 63.5G  0 part /persist

That means it can't be used to update to a slightly larger image.

Can't we fix that to use the same fixed size for the partitions?

eriknordmark reviewed (changes requested):
I cloned this workspace and tried what works on master, and it fails to boot the installer.

My invocation is:
make -e QEMU_MEMORY=16384 MEDIA_SIZE=65536 SSH_PORT=2345 installer-raw run-installer-raw run-target

rouming commented:
@eriknordmark 
> That means it can't be used to update to a slightly larger image.
> Can't we fix that to use the same fixed size for the partitions?

This is very good point, I forgot about upgrade. Actually without borrowing free space from the /persist we are screwed by default to upgrade to image greater than these hardcoded values. So in the future if image grows this has to be fixed by doing dangerous operations on the /persist. Now I can fix that by returning old values and taking max.

> I cloned this workspace and tried what works on master, and it fails to boot the installer.

This was fixed just before your comment, here https://github.com/lf-edge/eve/compare/8ab4fde0385564de8f57a560a255a6d8bed92179..21b71376ddaa6789ee9303e52b63c737743f2c9d
Please checkout once again.

rouming commented:
@eriknordmark @jsfakian 
Folks, now installer and verification images should work. I would appreciate if you offer me a hand in testing this.

Also the rootfs partition size is kept as is, i.e. 512Mb, so upgrade won't fail. This is for the imga, imgb partitions only, the installer/verificator keeps rootfs as a file on fat32, so we don't keep this max value.

rouming commented:
> > Another great feature becomes possible is to "patch" resulting qemu image with changed files and not generate rootfs from scratch, which can take minutes of CPU time
> 
> How does that work?

@deitch here #3842. Can we continue improving? Still a lot of room for improvements, but help of the linuxkit is required.

deitch commented:
This needs to be approved and merged before #3842

deitch commented:
> Also the rootfs partition size is kept as is, i.e. 512Mb, so upgrade won't fail. This is for the imga, imgb partitions only, the installer/verificator keeps rootfs as a file on fat32, so we don't keep this max value.

So this is ready to go?

> Still a lot of room for improvements, but help of the linuxkit is required

For this PR? Or for next steps you want in #3842? Or for something else beyond that?

rouming commented:
> > Also the rootfs partition size is kept as is, i.e. 512Mb, so upgrade won't fail. This is for the imga, imgb partitions only, the installer/verificator keeps rootfs as a file on fat32, so we don't keep this max value.
> 
> So this is ready to go?
> 
> > Still a lot of room for improvements, but help of the linuxkit is required
> 
> For this PR? Or for next steps you want in #3842? Or for something else beyond that?

This is ready to go. There was some important concern from @eriknordmark , I fixed that. If he does not have any objections  or anyone else, this can be merged. Also would be great if @jsfakian can help me with verification image. Last time he spotted one thing, which breaks the image. This is also fixed.

deitch commented:
Yeah, I see @eriknordmark still has changes pending. Let's let him sign off and merge it in. Hats off @rouming

eriknordmark reviewed (approved):
LGTM but i haven't had time to retry my make test
